### ISSUE TITLE
[Text] How to show "Read more" when using numberOfLines property?

### ISSUE THREAD

**Issue by ranyefet** [facebook/react-native#2496]
Hi all,

I have a long text with `<Text>` element and I use  `numberOfLines` to truncate the text and show "..." at the end.

I also want to show `Read More` link If the text was truncated and while clicking on it will show the full text.

How can I know if the text was truncated or not?

Thanks!

**Comment by brentvatne** [comment]
Hi there!

There's no way to know if it has been truncated, maybe it would be nice to have that option available although I'm not sure what the best way to expose it would be.

```
// Non-existent API that might work
<Text numberOfLines={3} renderTruncatedFooter={() => <ReadMoreButton />}>Some text</Text>
```

For now, you could just truncate the text manually when it exceeds some number of characters that you define, and add the the `Read more` button manually.

cc @ide

**Comment by ranyefet** [comment]
@brentvatne Thanks for the comment! I sure be useful to have that. 
I found [React Elipsify](https://bebraw.github.io/react-ellipsify/) which looks great but only work with HTML. It's too bad that we have duplicate components only because the rendering target is different

Is there a standard way to share components between React for Web and React Native?
Maybe have <View> & <Text> that will translate to <div> & <span> ?

We shou

**Comment by brentvatne** [comment]
@ranyefet - alas, there is no standard way to do that right now :(

**Comment by aleclarson** [comment]
I'm interested in seeing `renderTruncatedFooter` get implemented. :smile: 

cc @ide

**Comment by sinjimmyi** [comment]
I got this worked by call the [measure](https://facebook.github.io/react-native/docs/nativemethodsmixin.html) function. At first set the numberOfLines=0 , then in the measure funtion, I can get the height of the view, if the value of height is bigger than my setting value(in my case 100 pt for 4 lines)。it means there are some word need to truncate，so I set numberOfLines to 4,and show the show more button.

**Comment by aleclarson** [comment]
Ideally, you should avoid manual text layout. But it's sometimes required. Thus, I implemented some helper functions on `NativeModules.TextManager`.

How you use these helpers to achieve truncation is up to you. But I will eventually report back here with details about doing so.

---

Use `TextManager.computeWidth()` for **single-line** `Text` components.

[ See the implementation](https://github.com/aleclarson/react-native/blob/v0.19.0/Libraries/Text/RCTTextManager.m#L66-L85).

```js
TextManager.computeWidth("any string", {
  fontFamily: String,
  fontSize: Number,
  fontWeight: Number,
  fontStyle: String,
  letterSpacing: Number,
}, function(width) {
  // The returned number represents the total width of the given string.
});
```
---

Use `TextManager.computeLines()` for **multi-line** `Text` components.

[See the implementation](https://github.com/aleclarson/react-native/blob/v0.19.0/Libraries/Text/RCTTextManager.m#L87-L135).

```js
TextManager.computeLines("any string", {
  width: Number,
  fontFamily: String,
  fontSize: Number,
  fontWeight: Number,
  fontStyle: String,
  letterSpacing: Number,
  lineBreakMode: String, // This needs other commits to be supported.
}, function(lines) {
  // The returned array contains each line of text.
});
```

**Comment by jsierles** [comment]
@facebook-github-bot stack-overflow

**Comment by brentvatne** [comment]
@aleclarson - that looks neat!

**Comment by fungilation** [comment]
Especially when I'm not seeing an obvious way to animate Text view's numberOfLines, for a "Read more..." animated expansion on tap. The best I can do would be hiding overflowed text in a hidden second Text view, and show/expand that (with animation) on tapping "..." in the initial truncated Text block. But then I end up with 2 discontinuous text blocks, which is not ideal.

I propose adding a "isTruncated" prop (bool) to <Text>, and a complementary function for animating numberOfLines from X to Y. If Y>X it'd be a text expand, if not it'd be a collapse.

**Comment by aleclarson** [comment]
@fungilation I went ahead and posted your feature request on [Product Pains](https://productpains.com/post/react-native/make-textnumberoflines-animatable). :+1:

**Comment by fantattitude** [comment]
@aleclarson Your implementation links don't work anymore. Could you provide them again, I'm very interested.

**Comment by aleclarson** [comment]
@fantattitude Fixed! 👍

**Comment by fantattitude** [comment]
@aleclarson Do you think it'd be possible to add the method to a category (then not having to change the react-native I use to a fork I'd create) ?

**Comment by aleclarson** [comment]
@fantattitude I don't see why not, but I haven't tried. 😄

**Comment by fantattitude** [comment]
@aleclarson I can confirm, it works, just had to add Header Search Path to `RCTTextManager`. But now I'm stuck because my render happens before computeLines finish and changes my `state` even though I do it in `componentDidMount`…

**Comment by aleclarson** [comment]
@fantattitude You'll want to hide the element until `computeLines` finishes. Or you could call `computeLines` in a container component; then render the dependent component after `computeLines` finishes.

**Comment by leecade** [comment]
👍

**Comment by musicode** [comment]
how to resolve it?

**Comment by brentvatne** [comment]
made this lib based off of the approach of initially rendering with unbounded number of lines, then adjusting on the next frame and comparing height. check out the source, it's pretty straightforward.

https://github.com/exponentjs/react-native-read-more


![](https://raw.githubusercontent.com/exponentjs/react-native-read-more/master/example.gif)

**Comment by fungilation** [comment]
Looks perfect!

**Comment by sinjimmyi** [comment]
Great solution! Promise is more delicate than mine.

**Comment by Liroo** [comment]
Hey Brentvatne, looks like your solution is frame lost, we tried it on our application with react-native-parsed-text and it's not fast enough :/ 😭

**Comment by satya164** [comment]
@Liroo You could make the text with opacity 0 and then fade it in after measurements. Probably there should be a callback for that

**Comment by Liroo** [comment]
Yes sure, it's fast enough for one text maybe but we're using it in a timeline and there is potentially many texts to truncate, we have a frame with full text, a frame with truncated text and a frame with readMore button, it's really disturbing

**Comment by nlt2390** [comment]
I wrote this plugin to display show more/less. Basically, I set ` numberOfLines=null`, got the Height, then set back to provided number to know if the text was cut off or not.

[https://github.com/nlt2390/react-native-view-more-text](https://github.com/nlt2390/react-native-view-more-text)

**Comment by brentvatne** [comment]
> Yes sure, it's fast enough for one text maybe but we're using it in a timeline and there is potentially many texts to truncate, we have a frame with full text, a frame with truncated text and a frame with readMore button, it's really disturbing

Sounds like you need a native solution for this then! Agreed that the solution I posted isn't ideal for this use case

**Comment by christophermark** [comment]
Would it be helpful to expose a class method of `Text` for checking whether the text is currently being truncated or not? 

Android exposes a method that would allow us to check whether the text is being truncated. I could whip up a fix for Android if someone wants to tackle the iOS side!

**Comment by jck2** [comment]
@brentvatne I tried your solution and it works for most of what I need, but I was wondering if it's possible to put the read more button on the last line (instead of below the original text) in the event of truncation. Is that only possible right now if I truncate based on the number of characters?

**Comment by benvium** [comment]
@aleclarson Your methods don't seem to exist on RCTTextManager. Did they get pulled? Do we have any way of measuring the size of text?

**Comment by aleclarson** [comment]
@benvium Fixed the links on [my comment](https://github.com/facebook/react-native/issues/2496#issuecomment-191710224), again. Sorry about that. Those implementations are not currently in RN core.

**Comment by benvium** [comment]
@aleclarson Thank you. Looks like I can easily pull that implementation into a new plugin. I don't suppose you have an android implementation?

**Comment by aleclarson** [comment]
@benvium Nope, only working on iOS atm. Maybe someone else can contribute. 👍

**Comment by vvusts** [comment]
@brentvatne react-native-read-more doesn't work for state based text. If I change test state it doesn't trigger show/hide button :(

**Comment by brentvatne** [comment]
@vvusts - feel free to fork the project or submit a PR to add support for that!

**Comment by gameboyVito** [comment]
As @Liroo metioned, I am facing the same issue when I'm using the `expo/react-native-read-more-text`. In my case, I need to render a list with many items, each item contains a photo and a long text. Every time, when I refresh the FlatList or paginate it, I will see three steps to render: 

1.  a frame with full text
2. a frame with truncated text
3. a frame with readMore button

This three steps sometimes make the FlatList automatically scroll a few pixels or even a half screen. It is very annoying, I think the truncated text should be rendered before the component is shown up.

**Comment by sahrens** [comment]
@gameboyVito: it's not ideal, but if you don't want to use @aleclarson's solutions and don't have the experience to implement what you want as a custom native component, I would suggest doing the truncation based on the number of characters (we actually do this internally in a few places).

We should definitely solve this in the core though since it is a pretty common use-case. I think custom ellipses and ellipses onPress handler support would do it? We should also add text measurement functions because they are generally useful. Not sure when we would get to this though, but would appreciate PRs...

**Comment by Liqiankun** [comment]
renderTruncatedFooter？

**Comment by duhseekoh** [comment]
@Liqiankun what does this mean? That doesn't exist. Are you suggesting thats the name that it should be implemented under?

> renderTruncatedFooter?

**Comment by gameboyVito** [comment]
This feature must be done inside of NativeModule, instead of measure the size of text through JS. And, it should provide a API to control whether to show the truncated text or the full text.

**Comment by gasipari** [comment]
Speaking of NativeModule, does anyone know how to make a bridge between [GTMFadeTruncatingLabel](https://github.com/jonreid/google-toolbox-for-mac/blob/master/iPhone/GTMFadeTruncatingLabel.m) and RN's <Text/> ?

Reference: [Fade out end of text label strings that don't fit (instead of truncate)](https://stackoverflow.com/questions/12154313/fade-out-end-of-text-label-strings-that-dont-fit-instead-of-truncate)

Thanks!

**Comment by joelGarcia93** [comment]
as @brentvatne say, we can use this attribute numberOfLines={3} in the Text Component. for example..

<Text numberOfLines={3}>some text some text some text some text some text some text some text some text some text some text some text some text some text some text</Text

**Comment by Ilario17** [comment]
@gameboyVito Hi! I have the same problem, have you found a solution?

**Comment by GNarek** [comment]
@Ilario17 You may find solution for your case using tricks described here http://cmichel.io/how-to-get-the-size-of-a-react-native-view-dynamically/

**Comment by Ilario17** [comment]
@GNarek the problem is that I should add "Read more" in a Flatlist, So when you scroll down there ins't problem, but when you scroll up the recreated objects have a different height and then the table automatically shrinks giving a very bad feedback to the user

**Comment by sahrens** [comment]
Just store/fetch the data in a cache outside of the component, like redux or whatever you have for client state. Make sure the key you use is sufficiently unique that you won’t get collisions, but is reproducible when the component remounts.

**Comment by NaveenGoyal** [comment]
@brentvatne: Sometime calling the _handleTextReady Function taking time to called and because of that Read More button appears very late.
